(Default)Dict for some attributes

filchristou opened the pull request:
There was some talk about a newer interface taking advantage of the `utils.jl/getattr()` method.
https://github.com/MakieOrg/GraphMakie.jl/pull/67#discussion_r818534834

It happened that I needed it for some arguments and I developed this PR for a possible implementation.

I implemented it just for `nlabels_textsize` and `node_size`.
If you agree with the implementation it's rather straight forward to expand it for all other arguments.

### DefaultDict
I didn't really like the idea of bringing in a dependency for `DataStructures.jl`, but as much I thought about it, I couldn't find a way to leverage multiple dispatch to handle it transparently.
As a result I added one more check to the the `getattr()` method.
I think overall it is a nice feature though, to have an `DefaultDict` as an input.

### Example
```julia
julia> using Graphs, DataStructures
julia> gc = circular_ladder_graph(5)
julia> graphplot(gc; nlabels=repr.(vertices(gc)), nlabels_textsize=Dict(9=>0, 10=>0), node_size=Dict(9=>0, 10=>0))
julia> graphplot(gc; nlabels=repr.(vertices(gc)), nlabels_textsize=DefaultDict(30,Dict(9=>0, 10=>0)), node_size=DefaultDict(5,Dict(9=>0, 10=>0)))
```
Generating the following plots:
![image](https://user-images.githubusercontent.com/27723218/191175933-7e364ddc-5b3a-4289-8803-4f685120f766.png)
![image](https://user-images.githubusercontent.com/27723218/191176256-13eb830b-a4ee-4224-a3ae-400af81e1942.png)

### Conclusions
This PR is not meant to be merged without any discussion.
Instead, I wanted to suggest a way of implementing the new interface and talk it over.
Personally I believe that if we want to do exactly this for all attributes, the code could get messy.
So I think in the future it would be good to do all these new `lift`s inside a function or a macro.

hexaeder commented on src/utils.jl:
```diff
@@ -49,6 +49,8 @@ Else return the one and only element.
+    elseif o[] isa DefaultDict || o[] isa DefaultOrderedDict
+        return getindex(o[], idx)
     elseif o[] isa AbstractDict
         return get(o[], idx, default)
```
its a pity that `get(dict, key, default)` is not implemented for `DefaultDict` in a way that the "internal" default value is chosen over the "external" default defined in `get` call. Otherwise the `AbstractDict` codepath would just work

hexaeder replied:
```julia
julia> d = DefaultDict(:default, 1=>:foo, 2=>:bar)
DefaultDict{Int64, Symbol, Symbol} with 2 entries:
  2 => :bar
  1 => :foo

julia> get(d, 3, :wrongdefault)
:wrongdefault

julia> d[3]
:default

julia> get(d, 3, :wrongdefault)
:default
```
wat

filchristou replied:
Yes, `DefaultDict` works like that; after you query a key that doesn't exist, it then stores the default value in it.
I think the reasoning behind is that a default value could be a function, and in this case you could you it for memoization purposes. There is also an example here https://juliacollections.github.io/DataStructures.jl/latest/default_dict/

filchristou replied:
> its a pity that get(dict, key, default) is not implemented

Indeed a pity. that was also my perspective. Maybe we can start a discussion with the DataStructures.jl guys about another way `get(...)` could work but I wouldn't be too confident.

hexaeder commented:
Overall I am in favour of doing this. `get_attr` was added much later when i needed to handle edge parameters very differently depending on `edgeplot` using `linesegmens` or `beziersegments`. I never took the the time to properly expand the usage to all the arguments. Maybe it would be worth looking how it is done in Makie itself, the problem of distributing attributes which may be given in various formats to their individual recipients must be handled there all the time.

Other things to consider:
- Makie can handle observables of single values. I think it would be bad to `@lift` `nodesize=5` to an vector. This might be a problem bad for very large networks.
- The code should be extended to relift on size changes, but thats easily doable.

Just brainstroming here, but maybe we should do something like
```julia
prep_arguments(x, size, default) = x
function prep_arguments(x::Observable{AbstractDict}, size, default)
    @lift [collect based on $size and $x and $default]
end
function prep_arguments(x::Observable{DefaultDict}, size, _)
    @lift [collect based on $size and $x]
end
```
so that in the recipe the code becomes something like `markersize=prep_arguments(gp.nodesize, size, default)`.
But thats not good if the arguments come in as `Observable{Any}`...

hexaeder commented:
Sorry i totally forgot about that PR. And also I am an Idiot: i didn't like how it adds another dep, however, we already implicitly depend on `DataStructures` (via `Graphs` and others) so this really is a non-issue.

I am not a huge fan of the implementation of `DefaultDict` because it does not behave like I'd expect for `get`. But it is available and used so better to build on that then to cook up something new. So I'd say if you want to add this for more arguments go for it. The very least would be to document its usage for the parts where it is already implemented in that PR and add some tests.

filchristou commented:
you made some valuable comments [above](https://github.com/MakieOrg/GraphMakie.jl/pull/89#issuecomment-1251943219) and I didn't have time to process them in detail. Especially looking into the `Makie` package to see how they are doing it. So if you have more concrete examples of how `Makie` is doing it please let me know.

> So  I'd say if you want to add this for more arguments go for it.

I would prefer that we first decide on the structure of the code, merge this PR as a prototype, and then we can slowly start working on gradually doing the same for all attributes as a non-brainer.

I am already using this feature  locally, so sooner or later I will focus on finalizing & merging this PR..

filchristou commented:
for a demo you can try the following:
```julia
julia> using GraphMakie, GLMakie, Graphs;

julia> gc = circular_ladder_graph(5);

julia> ons = Observable(30);

julia> onf = Observable(30);

julia> graphplot(gc; nlabels=repr.(vertices(gc)), node_size=ons, nlabels_fontsize=onf)

julia> ons[] = 10; # check changes

julia> onf[] = 10; # check changes

julia> ons = Observable(Dict(1=>30, 10=>30));

julia> onf = Observable(Dict(1=>30, 10=>30));

julia> graphplot(gc; nlabels=repr.(vertices(gc)), node_size=ons, nlabels_fontsize=onf)

julia> onf[] = Dict(7=>30); # check changes

julia> ons[] = Dict(7=>30); # check changes

julia> ons = Observable(DefaultDict(70, 1=>30, 10=>30));

julia> onf = Observable(DefaultDict(70, 1=>30, 10=>30));

julia> graphplot(gc; nlabels=repr.(vertices(gc)), node_size=ons, nlabels_fontsize=onf)

julia> ons[] = DefaultDict(20, 10=>70); # check changes

julia> onf[] = DefaultDict(20, 10=>70); # check changes
```

filchristou commented:
>    Other things to consider:

- [x] Makie can handle observables of single values. I think it would be bad to @lift nodesize=5 to an vector. This might be a problem bad for very large networks.
- [x] The code should be extended to relift on size changes, but thats easily doable.

filchristou commented:
@hexaeder please let me know what do you think. As already mentioned this PR is a prototype and the other arguments should follow rather easily. The documentation and tests you mention should follow after we reach an agreement :)

hexaeder commented on src/recipes.jl:
```diff
@@ -331,6 +336,16 @@ function Makie.plot!(gp::GraphPlot)
+        end
+    end
+end
+
```
Should `prep_arguments` maybe forward arguments given as a plain vector rather than looping through them and creating a new one?

pros: prob faster
cons: not possible to denote default by using `nothing` inside vector (not sure if thats a feature^^); If we loop through it it is possible to pass a "longer" vector than needed. this might help with adjusting the graph size which is allways annoying.

filchristou replied:
> cons: not possible to denote default by using nothing inside vector 

when passing in a typeof `AbstractVector` it looks like there is no codepath to use any default value. Or is it ? Can you please elaborate ?

> this might help with adjusting the graph size

Yeah. I don't know how often people actually do it, but then it will be at least theoretically possible.

filchristou replied:
> this might help with adjusting the graph size

To take the best of both worlds probably probably we need to write more complicated logic. something like
```julia
    @lift begin
        if isscalar($o)
            return isnothing($o) ? $default : $o
        elseif $o isa AbstractVector
            length($indices) == length($o) && return $o
        end
        return [getattr(o, i, $default) for i in $indices]
    end
```

But currently I have the impression that will still break if the size of graph is changed, since in the `AbstractVector` case in `getattr` no expansion is done with any default value.
So basically this logic needs to enter in the `getattr`:
```julia
function getattr(x, idx, default=nothing)
    if x isa AbstractVector && !isa(x, Point)
        return idx < length(x) ? x[idx] : default
    elseif x isa DefaultDict || x isa DefaultOrderedDict
...
```

hexaeder replied:
On this pr, if i pass `v=[1, nothing, 3]` it will call `[getattr(e, i, default) for e in v]` which will replace `nothing` by the default of the recipe inside `getattr`. But I think this isnt really a feature. I think i'd vote for adding
```julia
elseif $o isa Vector
    $o
else
```

For adjusting the graph size using a default dict might be the best option anyway, i like that.

filchristou replied:
Which introduces a check in each iteration plus the allocation of new vector due to the comprehension. I am wondering whether we can avoid both by mutating the `AbastractVector` with `push!` ( or `deleteteat!` if graph decreases in size) but this will probably need even more complicated logic. 
Or something like `vcat(o$, fill(default, residualsize))` (or `deleteat!(o$, length(vertices(graph)):length($o)` for decreasing graph size).

filchristou replied:
But I am wondering are there use cases for changing the graph size ? I mean doing `add_vertex!` or `add_edge!` don't even trigger the Observable calls. One should substitute the graph as a whole.

hexaeder replied:
We have #92 which is about changing graph size. However switching out the graph is something more complicated and need attantion throughout the entire recipe and would be a job for another PR so lets not overcomplicate things here. I think i prefer the explicitness of passing a vector directly without meddling to much.
```
elsif $o isa AbstractVector
    o[]
else
```
will explicitly fail in both cases: `length(o) > nv(g)` and `lenght(o) < nv(g)`. I like that over the current state of the PR where it works if the passed vector is to long because we only loop through and basicially select a subset of the given vector.

hexaeder reviewed (commented):
nice, example works great!

filchristou commented on src/utils.jl:
```diff
@@ -48,13 +48,22 @@ Else return the one and only element.
+
+Return `true` if `Observable` is not an `AbstractVector` or an `AbstractDict`
+"""
+isscalar(o) = !isa(o, AbstractVector) && !isa(o, AbstractDict)
```
>Maybe the name isscalar is a bit misleading

Yes, naming it `issingleattribute` is better. 

> Do we have arguments where people might put a single `Point` ?

good catch. Yes there are the offsets that one can pass in a `Point`
So I guess we should update the function to 
```julia
issingleattribute(x) = isa(x, Point) || (!isa(x, AbstractVector) && !isa(x, AbstractDict))
```
(also substitute `o` with `x` since it's misleading that it might be an Observable while it's not)

filchristou commented:
Actually I also thought about a small change that in my opinion will lead to cleaner code (non-disruptive):

Change function `getattr(o::Observable, idx, default=nothing)` to
```julia
getattr(o::Observable, idx, default=nothing) = getattr(o[], idx, default
```
Add 

```julia
function getattr(x, idx, default=nothing)
    if x isa AbstractVector && !isa(x, Point)
        return x[idx]
    elseif x isa DefaultDict || x isa DefaultOrderedDict
        return getindex(x, idx)
    elseif x isa AbstractDict
        return get(x, idx, default)
    else
        return x === nothing ? default : x
    end
end
```

So then the loop can be
```julia
@lifft begin
...
[getattr($o, i, $default) for i in 1:$size]
...
end
```
intead of 
```julia
@lift begin
...
[getattr(o, i, $default) for i in 1:$size]
...
end
``` 
which I always find a bit awkward to handle observables as observables and not values inside a `@lift`.

What do you think ?

hexaeder commented:
Sure, you may redefine `getattr` that way. While you're at it: you can use the same `issingleargument` function ther since it means the same thing.

Have you looked into applying those things to the edge attributes yet? There it is a bit more complicated since `edgeplot` either calls `linesegmens` (which needs single attribute or vetor so we need to `prep_arguments`) or bezierpathes (which makes use of `getattr` anyway and can handle dicts and defaultdicts). So the `prep_arguments` shoult probably be applied inside the `linesegments` path, but i havn't really thought this through.

hexaeder commented:
Also, since `getattr`, `prep_arguments` and `issingleattribute` are all part of the same processing i think they should live next to each other in the codebase. I don't care whether thats recipe or utils. Also, mache `prep_arguments` should be called `prep_attributes` or something?

filchristou commented:
I see. No I still haven't looked into edges.

> prep_arguments should be called prep_attributes or something?

yes, sound better.

filchristou commented:
Since we started this work, I think it's better we might as well finish off all attributes in this PR.
This way also docs/tests can be made more complete.

Following a list of the attributes and which ones can support the new feature:
- [ ]         layout = Spring(),

node attributes (Scatter)

- [x]         node_color = scatter_theme.color,
- [x]         node_size = scatter_theme.markersize,
- [x]         node_marker = scatter_theme.marker,
- [ ]         node_attr = (;),

edge attributes (LineSegements)
- [x]         edge_color = lineseg_theme.color,
- [x]         edge_width = lineseg_theme.linewidth,
- [ ]         edge_attr = (;),

arrow attributes (Scatter)
- [ ]         arrow_show = automatic,
- [ ]         arrow_marker = '➤',
- [ ]         arrow_size = scatter_theme.markersize,
- [ ]         arrow_shift = 0.5,
- [ ]         arrow_attr = (;),

node label attributes (Text)
- [ ]         nlabels = nothing,
- [x]         nlabels_align = (:left, :bottom),
- [x]         nlabels_distance = 0.0,
- [x]         nlabels_color = labels_theme.color,
- [x]         nlabels_offset = nothing,
- [x]         nlabels_fontsize = labels_theme.fontsize,
- [ ]         nlabels_attr = (;),

edge label attributes (Text)
- [ ]         elabels = nothing,
- [x]         elabels_align = (:center, :center),
- [x]         elabels_side = :left,
- [x]         elabels_distance = automatic,
- [x]         elabels_shift = 0.5,
- [x]         elabels_rotation = automatic,
- [x]         elabels_offset = nothing,
- [x]         elabels_color = labels_theme.color,
- [x]         elabels_fontsize = labels_theme.fontsize,
- [ ]         elabels_attr = (;),

self edge attributes (already use `getattr` so no need ? )
- [ ]         edge_plottype = automatic,
- [x]         selfedge_size = automatic,
- [x]         selfedge_direction = automatic,
- [x]         selfedge_width = automatic,
- [ ]         curve_distance = 0.1,
- [ ]         curve_distance_usage = automatic,
- [ ]         tangents=nothing,
- [ ]         tfactor=0.6,
- [x]         waypoints=nothing,
- [x]         waypoint_radius=nothing,

filchristou commented:
I start to be happy with where this goes and I think we are close to merging. 
Have a look at my remarks here where I make a summary and I highlight the new additions.
I took some liberty on some things so feel free to be critical about them.

# Summary
Mainly, this PR extends functionality by supporting arguments of `Dict`.
However there are some other ideas integrated.

### Bug Fixes
Testing my previous workflow https://github.com/MakieOrg/GraphMakie.jl/pull/89#issuecomment-1450975745
there were some errors. More specifically updating the `Observable` properties errored. I fixed the bug in the following lines
- https://github.com/MakieOrg/GraphMakie.jl/pull/89/files#diff-8861ed4224a2baebdb2c6efa750e27a9d73b2b84f72f4f46b153786969fd7f8eL259
- https://github.com/MakieOrg/GraphMakie.jl/pull/89/files#diff-8861ed4224a2baebdb2c6efa750e27a9d73b2b84f72f4f46b153786969fd7f8eL304
- https://github.com/MakieOrg/GraphMakie.jl/pull/89/files#diff-8861ed4224a2baebdb2c6efa750e27a9d73b2b84f72f4f46b153786969fd7f8eR747
The problem was that in the `@lift` command not the `Observable` type was used.

### Renaming variables
In lines like https://github.com/MakieOrg/GraphMakie.jl/blob/7c2e706a37d29307ca2b0b1cc8dc9d42691d6da2/src/recipes.jl#L206 , https://github.com/MakieOrg/GraphMakie.jl/blob/7c2e706a37d29307ca2b0b1cc8dc9d42691d6da2/src/recipes.jl#L235 ,
https://github.com/MakieOrg/GraphMakie.jl/blob/7c2e706a37d29307ca2b0b1cc8dc9d42691d6da2/src/recipes.jl#L286,
the same name as the recipe Observable is given to represent same reference.
This was in contrast to how [`node_color`](https://github.com/MakieOrg/GraphMakie.jl/pull/89/files#diff-8861ed4224a2baebdb2c6efa750e27a9d73b2b84f72f4f46b153786969fd7f8eL262), [`node_marker`](https://github.com/MakieOrg/GraphMakie.jl/pull/89/files#diff-8861ed4224a2baebdb2c6efa750e27a9d73b2b84f72f4f46b153786969fd7f8eL268), [`node_strokewidth`](https://github.com/MakieOrg/GraphMakie.jl/pull/89/files#diff-8861ed4224a2baebdb2c6efa750e27a9d73b2b84f72f4f46b153786969fd7f8eL274) are meant which are different dependent `Observable`. I think it's more clear to give them different name. I opted for `name_m` where "m" stood for modified.

### Extension of Dict to all supporting arguments
The main work of this PR is modifying `get_attr` and adding the `prep_arguments` functions as already discussed some months ago. Now, I just inlined the call to this function to all appropriate recipes.

### Use of Dict{Edge} for edge properties
We 've talked in the past about supporting `Dict{Edge}` for edge properties.
At the beginning I was hesitant to implement something like that since for big graphs it could get slow and memory hungry.
However I think with the current design it's okey.
Already at compile time (or one dynamic dispatch per `graphplot`) the correct flavor of `getedgeinds` is called that returns the appropriate indexing. Only when the graph is undirected and an `AbstractDict{<:AbstractEdge}` is given there is some further computation, which is being done on demand using `Iterators` rather than saving it in memory.
To further increase efficiency we can use some lazy implementation that caches the result.

### SimpleTraits.jl dependency
`SimpleTraits.jl` is already in the `Manifest.toml` as it's a dependency of `Graphs.jl`. 
I find it elegant to use it also here.

### Docs 
I updated the docs to talk about the new feature.
I don't have examples there but I reference the reftests

### Reftests
I introduced quite some reftests.
I wanted to introduce them somewhere in the middle of the document as they would be better categorized, but I realized I had to change all the assets numbering. Is there a clever way to do that, or should we always add reftests in the end to avoid such developer overheads ?

### Versioning
This is non-breaking but still introduces some obvious new features. Should that be a minor or patch update ?

### docs/Manifest.toml
The `./docs` should depend on the GraphMakie as `dev ..`. Otherwise it wouldn't follow locally up until the version was published.
Am I right ? How do you develop locally when there are such changes ?
If you like, this could be a separate PR.

# Remarks
I just wanted to take the opportunity to highlight some points

### element-wise `automatic`
Currently the Dict funcionality can have a single default value.
However, some arguments like label rotation etc. have different default values per element.
I think we should support that the fallback for `Dict` and `DefaultDict` should be these default values.
However, I would advise that to be a separate PR.

### Need for benchmarks
As I mentioned I was a bit concerned with performance and currently there is no out-of-the-box way to test how we do.
There are also some other julia packages we can compare to, or even to the python ecosystem.
This can bring in a new dimension to this package, which I think is desirable.

hexaeder commented on src/recipes.jl:
```diff
@@ -287,42 +289,43 @@ function Makie.plot!(gp::GraphPlot)
-                          color=gp.edge_color,
-                          linewidth=gp.edge_width,
-                          gp.edge_attr...)
+        color=prep_attributes(gp.edge_color, @lift(getedgeinds($graph, $(gp.edge_color))), dfth.edge_color),
```
Funcionally I am on board but I am not super happy with this construct. Because a change of `gp.edge_color` from the outside would trigger two things:
- recalculation of observable definition within `prep_attributes`
- recalculation of `getedgeinsd` output which then triggers recalculation of `prep_attributes`

Havn't though this through, but maybe it is worth splitting `prep_attributes` in 
```
prep_node_attributes(attr::Observable, graph::Observable{g}, default)
``` 
and 
```
prep_edge_attributes(attr::Observable, graph::Observable{g}, default)
    @lift begin
        if issingleattribute($attr)
            isnothing($attr) ? $default : $attr
        elseif $attr isa AbstractVector
            $attr
        else
            [getattr($attr, i, $default) for i in getedgeinds($graph, $attr)]
        end
    end
end
```
So `@lift getedgeinds(..)` for edges and `@lift vertice(g..)` could become part of the `prep_attributes` lift.

hexaeder replied:
There are two subproblems:
- even if `gp.edge_color` is an vector, the `tempobs` will be updated every single time
- the observable created in `prep_attributes` lifts both on `gp.edge_color` and on the `tempobs`. Therefore, change of `gp.edge_color` will trigger `prep_attributes`, then `tempobs` and then `tempobs` will trigger `prep_attributes` again.

I think this function
```julia
prep_edge_attributes(attr::Observable, graph::Observable{g}, default)
    @lift begin
        if issingleattribute($attr)
            isnothing($attr) ? $default : $attr
        elseif $attr isa AbstractVector
            $attr
        else
            [getattr($attr, i, $default) for i in getedgeinds($graph, $attr)]
        end
    end
end
```
elides both of those problems. I think withint `prep_edge_attributes` it still makes perfect sense to use the trait dispatches function `getedgeinds`. But we get rid of one layer of observables.

filchristou replied:
You are right! Nice catch! I though there was a dependency tree that decided what/how to execute with `Observables`.
However the following code which is similar indeed is executed twice:
```julia
using Observables

edge_color = Observable(1)
graph = Observable(2)

tempobs = map(edge_color, graph) do e,g
    @info("tempobs update")
    e + g
end

prep_attrs = map(edge_color, tempobs) do e, t
    @info("prep_attrs update")
    e + t
end

myplot = map(prep_attrs) do p
    @info("myplot update")
    p * 10
end
```

Changing edge_color[] will execute prepr_attrs twice:
```julia-repl
julia> edge_color[] = 3;
[ Info: tempobs update
[ Info: prep_attrs update
[ Info: myplot update
[ Info: prep_attrs update
[ Info: myplot update
```

I will rewrite it as you suggested together with the other comments. 
Also in order to keep it symmetric I would also make a `prep_vertex_attributes` that also gets a graph `Observable`.

hexaeder commented on src/utils.jl:
```diff
@@ -38,23 +38,85 @@ function _get_label_plot(gp::GraphPlot, labels)
+
+Return enumeration of edges for directed graph
+"""
+@traitfn function getedgeinds(gr::G, edgedat::D) where {G<:AbstractGraph, K<:AbstractEdge, D<:AbstractDict{K}; IsDirected{G}}
```
i think `getedgekeys` would be more fitting since indices are allways Integers?

filchristou replied:
ok

hexaeder commented on src/utils.jl:
```diff
@@ -38,23 +38,85 @@ function _get_label_plot(gp::GraphPlot, labels)
+Return enumeration of edges for undirected graph such that the user's keys are used
+"""
+@traitfn function getedgeinds(gr::G, edgedat::D) where {G<:AbstractGraph, K<:AbstractEdge, D<:AbstractDict{K}; !IsDirected{G}}
+    Iterators.map(e -> reverse(e) ∈ keys(edgedat) ? reverse(e) : e , edges(gr))
```
I don't get that part. Wouldn't that mean that a graph
```
g = SimpleDiGraph(2)
add_edge!(g, 2, 1)
```
could be colored using
```
ecolor = Dict(SimpleEdge(1,2)=>:red)
```
just because `1->2` happens to be present in the provided dict? Is that desired? I mean in directed graphs `1->2` and `2->1` are just different objects and i would expect to hit the default path then.

filchristou replied:
Currently your example will produce a black edge as expected
```julia
graphplot(g, edge_color=Dict(Edge(2,1)=>:red))
```
![image](https://github.com/MakieOrg/GraphMakie.jl/assets/27723218/3386240d-e223-4df0-937a-f9dec1819d5e)

You can think about it like that: In case of undirected graphs we modify the `edges()` function such that it uses the same naming as the user used in the provided `Dict`. So for every element in `edges()` it will use either the one as the user identified the specific edge or the default from the `edges()`

For undirected graphs always the [`IsDirected`](https://github.com/MakieOrg/GraphMakie.jl/pull/89/files#diff-47c27891e951c8cd946b850dc2df31082624afdf57446c21cb6992f5f4b74aa2R46) function is called which always returns exactly `edges()`

I still think it works as intended. 
Would you agree now ? :)

I can add an extended help doc for the developers, as for the user it should be intuitive.

hexaeder replied:
Hah, sorry for that. I didn't pay enought attantion. That codepath is hit in case of **undirected** graphs while I was thinking it is hit in case of **directed** graphs where it doesn't make sense. The differens a `!` can make...

hexaeder commented:
>  I think it's more clear to give them different name. I opted for name_m where "m" stood for modified.

totally agree

> Use of Dict{Edge} for edge properties

see comments on files, i think we can reduce some unecssary updates by splitting prep arguments in two functions for nodes and edges

> I wanted to introduce them somewhere in the middle of the document as they would be better categorized, but I realized I had to change all the assets numbering. Is there a clever way to do that, or should we always add reftests in the end to avoid such developer overheads?

No, not yet. There is a bit of a tradeoff, currently we only need to annotate with `@reftest` and number them automaticially by file name. I have an not-yet-registered package [`PlotReferenceTests.jl`](https://github.com/hexaeder/PlotReferenceTests.jl) which does things similar but slightly different, there you'd need to give every reftest a name `@reftest fig "example"`. I think it makes sense to switch eventually as it also works much better when interactivly stepping through those tests out of order (as I often do during debugging).

> Versioning

I see that it is a big feature upgrade but I'd still opt for non-breaking release. I think there is no reason to make everybody bump there versions.

> docs/Manifest.toml

What i normally do is executing `docs/localmake.jl` from a terminal. It will set up the doc env similar to what happens in CI. And if you have `LiveServer` in your global env it will serve the docs and you can repeatadly rebuild them without restarting the julia session. Even though I just noticed that there should be a `using Revise` in there too.

> element-wise `automatic`

I agree, this changes the structure quit a bit but is relatively simple and should be done in a separate PR. Basicially we would first need to expand/prep the arguments  before resolving the automatics, which is currently done first creating those `_m` version.

> benchmarks

Agree. I think there is room for a lot of improvement in the package, currently everything is super dynamic and type unstable.

hexaeder commented:
I'll merge this into a feature branch of the repo to get the doc preview build, you should have all the rights to push to that new branch too
